Use libepoxy

yshui opened the pull request.

yshui committed:
build: add libepoxy

Add libepoxy dependency to CI manifest and Nix.

For Nix, we need to set shellHook to workaround a NixOS limitation, see:

https://github.com/NixOS/nixpkgs/issues/287763

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
backend: gl: use libepoxy's has_*_extension

So we don't need maintain our own version.

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

yshui committed:
Update README.md and CHANGELOG.md

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

absolutelynothelix commented on CHANGELOG.md:
```diff
@@ -4,9 +4,14 @@
 ## Bug fixes
 
 * Workaround a NVIDIA problem that causes high CPU usage after suspend/resume (#1172, #1168)
+* Fix building on OpenBSD (#1189 #1188)
```
```suggestion
* Fix building on OpenBSD (#1189, #1188)
```

absolutelynothelix commented on flake.nix:
```diff
@@ -24,16 +24,20 @@
       hardeningDisable = [ "fortify" ];
+      shellHook = ''
+        export LD_LIBRARY_PATH+=":/run/opengl-driver/lib"
+      '';
```
afaiu it's a workaround and i think i saw it noted in one of the commits' messages. maybe it's worth a mention here as well if comments are allowed in flakes?

absolutelynothelix commented on src/backend/gl/egl.c:
```diff
@@ -422,66 +395,22 @@ struct backend_operations egl_ops = {
-#define check_ext(name) eglext.has_##name = egl_has_extension(dpy, #name)
+#define check_ext(name)                                                                  \
+	eglext.has_##name = epoxy_has_egl_extension(dpy, #name);                         \
+	log_info("Extension %s - %s", #name, eglext.has_##name ? "present" : "absent");
```
honestly, i don't like the format of this message but i can't think of anything really better right now.

absolutelynothelix commented on src/backend/gl/glx.c:
```diff
@@ -545,62 +545,17 @@ struct backend_operations glx_ops = {
-#define check_ext(name) glxext.has_##name = glx_has_extension(dpy, screen, #name)
+#define check_ext(name)                                                                  \
+	glxext.has_##name = epoxy_has_glx_extension(dpy, screen, #name);                 \
+	log_info("Extension " #name " - %s", glxext.has_##name ? "present" : "absent")
```
same as above.

absolutelynothelix replied:
also, this line is a bit different compared to the one above. let's make them the same at least (don't miss the semicolon at the end!).

yshui replied:
no i need to remove the semicolon on the other line.

absolutelynothelix commented on src/meson.build:
```diff
@@ -23,7 +23,7 @@ required_xcb_packages = [
 required_packages = [
 	'pixman-1', 'x11', 'x11-xcb', 'xcb-image', 'xcb-renderutil', 'xcb-util',
-	'xext'
+	'xext',
```
huh?

yshui replied:
trailing commas are good

absolutelynothelix commented on src/meson.build:
```diff
@@ -59,7 +59,10 @@ endif
 
 if get_option('opengl')
 	cflags += ['-DCONFIG_OPENGL', '-DGL_GLEXT_PROTOTYPES']
```
do we still need `-DGL_GLEXT_PROTOTYPES`?

yshui replied:
good catch

absolutelynothelix commented on src/meson.build:
```diff
@@ -59,7 +59,10 @@ endif
 
 if get_option('opengl')
 	cflags += ['-DCONFIG_OPENGL', '-DGL_GLEXT_PROTOTYPES']
-	deps += [dependency('gl', required: true), dependency('egl', required: true), dependency('threads', required:true)]
```
do we still explicitly depend on the `libgl` and `libegl`?

absolutelynothelix reviewed (commented):
looks good, builds and works for me. a few things may need to be addressed before merging.

yukiteruamano commented:
@yshui @absolutelynothelix I'm going to try the changes in OpenBSD, there shouldn't be any problems with using libepoxy, but it's better to clear up any doubts

yshui committed:
backend: gl: use libepoxy

There is actually no specification what symbols are exported from a
libGL implementation. The (extremely outdated) OpenGL ABI specification
says only GL 1.2 functions are guaranteed. Don't know how relevant that
is now, but different libGL implementations do export different set of
symbols. On Linux we are most likely to be linked with libglvnd, which
has everything we need. But on other platforms this is not necessarily
the case, for example on OpenBSD we are missing glGetQueryObjectui64v.

Use libepoxy so we can outsource this problem and never worry about it
ever again. Plus it also saves us from calling GetProcAddress ourselves.

Changes other than trivial build fixes I have to make:

1. Can't use eglCreatePlatformWindowSurface/eglGetPlatformDisplay.
   libepoxy checks for EGL 1.5 when resolving these functions. But
   without a current context, libepoxy assumes we only have EGL 1.4.
   This creates a chicken and egg problem - we need a display to call
   eglGetPlatformDisplay. We have to use the *EXT version instead.

Signed-off-by: Yuxuan Shui <yshuiv7@gmail.com>

absolutelynothelix commented:
@yukiteruamano, we now have an openbsd ci so picom should at least always build there. sorry for the inconvenience!

yukiteruamano commented:
Yeah, I see that and I test the last changes, compile without issues
